File: src/main/java/leetcode/problems/medium/Clone_Graph_133.java
package leetcode.problems.medium;
import leetcode.problems.util.Node;

import java.util.*;

public class Clone_Graph_133 {

    public Node cloneGraph(Node node) {
        if (node == null) return null;
        return cloneGraph(node, new HashMap<>());
    }

    Node cloneGraph(Node node, Map<Integer, Node> map) {
        Node root = new Node(node.val);
        map.putIfAbsent(node.val, root);
        for (Node neighbor : node.neighbors) {
            if (!map.containsKey(neighbor.val)) {
                root.neighbors.add(cloneGraph(neighbor, map));
            } else {
                root.neighbors.add(map.get(neighbor.val));
            }
        }
        return root;
    }

    public static void main(String[] args) {
        Clone_Graph_133 x = new Clone_Graph_133();
        Node n1 = new Node(1);
        Node n2 = new Node(2);
        n1.neighbors.add(n2);
        n2.neighbors.add(n1);
        Node nx = x.cloneGraph(n1);
        System.out.println("end.");
        //System.out.println(x.cloneGraph(n1));
    }


}

File: src/main/python/leetcode/medium/Clone_Graph_133.py
# 133. Clone Graph
# https://leetcode.com/problems/clone-graph/
#
#
#
#
from leetcode.Node import Node
from typing import Set
class Solution:

    def cloneGraph(self, node: 'Node') -> 'Node':
        return self.cloneGraph2(node, set())

    def cloneGraph2(self, node: 'Node', stack: Set[Node]) -> 'Node':
        root = Node(val = node.val)
        for neighbor in node.neighbors:
            if not neighbor in stack:
                stack.add(root)
                root.neighbors.append(self.cloneGraph2(neighbor, stack))
            else:
                root.neighbors.append(neighbor)
        return root

if __name__ == '__main__':
    x = Solution()
    n1,n2,n3=Node(1),Node(2),Node(3)
    n1.neighbors=[n2,n3]
    n2.neighbors=[n1,n3]
    n3.neighbors=[n2,n1]
    print(x.cloneGraph(n1))